lib/ddtrace/workers/runtime_metrics.rb

```diff
@@ -43,6 +43,17 @@ def associate_with_span(*args)
         metrics.associate_with_span(*args).tap { perform }
       end
 
+      # TODO: `close_metrics` is only needed because
+      # Datadog::Components directly manipulates the lifecycle of
+      # Runtime::Metrics.statsd instances.
+      # This should be avoided, as it prevents this class from
+      # ensuring correct resource decommission of its internal
+      # dependencies.
+      def stop(*args, close_metrics: true)
+        @metrics.close if close_metrics
+        super(*args)
+      end
```

Comment:
Is this safe? We're calling `.close` on something that is shared with a background thread, so if `Runtime::Metrics` or `Datadog::Statsd` are not thread-safe, this can be problematic.

I was thinking that perhaps we can have an easy fix by moving the close *after* the call to super -- then we know the background activity has stopped, so it should be a safer time to close the resource?

Reply:
I don't have a strong opinion here.
It's rather hard to reason about this worker class, as there are a few layers of pooling/threading/loop that interweave with each other... I always get lost in the process. It seems like something we'd want to "flatten" in the future.

I'll give it a shot and see if works correctly with close after super.

Reply:
Yes please! I think the module approach right now makes the combined objects have a lot of behavior and weird cases. I'd really like to move to using composition for these workers, e.g. two objects which communicate with clear APIs, rather than the current approach of having a lot of behavior injected into the class via a combination of include and prepend.